src/crypto/Address.dto.js

```diff
@@ -0,0 +1,97 @@
+// @flow
+
+import {
+  Address,
+  ByronAddress,
+  BaseAddress,
+  // PointerAddress,
+  // EnterpriseAddress,
+  RewardAddress,
+} from '@emurgo/react-native-haskell-shelley'
+
+// NOTE: ** IMPORTANT ** The order matters
+const ADDRESS_TYPES = Object.freeze([
+  ByronAddress,
+  BaseAddress,
+  // PointerAddress,
+  // EnterpriseAddress,
+  RewardAddress,
+])
+
+export type KeyHashesCardano = {|
+  spending: string,
+  staking: string,
+|}
+
+/**
+ * @description The resolver for the Cardano key hashes
+ * @param {any} wasmAddress A Cardano wallet Address
+ * @returns {Promise<KeyHashesCardano|null>} You don't need to test for Byron it will return null
+ */
+export const getKeyHashesCardano = async (
+  wasmAddress: any,
+): Promise<KeyHashesCardano | null> => {
+  let addr, pay, stake, paymentKeyHash, stakingKeyHash
+  for (const [i, addressType] of ADDRESS_TYPES.entries()) {
+    // First = Byron
+    if (!i) {
+      addr = await addressType.from_address(wasmAddress)
+      if (addr) return null
+    } else {
+      addr = await addressType.from_address(wasmAddress)
+      if (addr) {
+        pay = await addr.payment_cred()
+        stake = await addr.stake_cred()
+        if (!pay || !stake) break
+        paymentKeyHash = await pay.to_keyhash()
+        stakingKeyHash = await stake.to_keyhash()
+        if (!paymentKeyHash || !stakingKeyHash) break
+        return {
+          spending: Buffer.from(await paymentKeyHash.to_bytes())?.toString(
+            'hex',
+          ),
+          staking: Buffer.from(await stakingKeyHash.to_bytes())?.toString(
+            'hex',
+          ),
+        }
+      }
+    }
+  }
+  // Something went wrong
+  throw new Error('getSpendingKeyHash unknown address type')
+}
+
+export interface YoroiAddressInfoInterface<KeyHashes> {
+  +address: string;
+  +getKeyHashes: () => Promise<KeyHashes | null>;
+}
+
+// prettier-ignore
+export class AddressDTOCardano
```

Comment:
Note that there are plans to integrate Ergo so in general we try to organise the code keeping in mind that other wallet types and networks might be added in the future. In this case I'd have probably defined the interface inside `crypto` but all the ADA-specific logic inside `crypto/shelley/`. That folder is for logic specific to Shelley wallets from the "Haskell Shelley" era.

Reply:
Totally agree.